sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,481 @@
+                clip_value=clip_value,
+                gamma=gamma,
+                n_particles=n_particles,
+                **kwargs,
+            )
+
+        if (
+            reset_optimizer
+            or self._optimizer is None
+            or not isinstance(self._optimizer, self._optimizer_base)
+        ):
+            self._optimizer = self._optimizer_base(
+                self.potential_fn,
+                self.q,
+                lr=learning_rate,
+                clip_value=clip_value,
+                gamma=gamma,
+                n_particles=n_particles,
+                **kwargs,
+            )
+
+        # Check context
+        x = atleast_2d_float32_tensor(self._x_else_default_x(x)).to(self._device)
+        already_trained = self._trained_on is not None and (x == self._trained_on).all()
+
+        # Optimize
+        self._optimizer.update({**locals(), **kwargs})
+        optimizer = self._optimizer
+        optimizer.to(self._device)
+        optimizer.reset_loss_stats()
+
+        if show_progress_bar:
+            iters = tqdm(range(max_num_iters))
+        else:
+            iters = range(max_num_iters)
+
+        # Warmup before training
+        if reset_optimizer or (not optimizer.warm_up_was_done and not already_trained):
+            if show_progress_bar:
+                iters.set_description("Warmup phase, this takes some seconds...")
+            optimizer.warm_up(warm_up_rounds)
+
+        for i in iters:
+            optimizer.step(x)
+            mean_loss, std_loss = optimizer.get_loss_stats()
+            # Update progress bar
+            if show_progress_bar:
+                iters.set_description(
+                    f"Loss: {np.round(mean_loss, 2)} Std: {np.round(std_loss, 2)}"
+                )
+            # Check for convergence
+            if check_for_convergence and i > min_num_iters:
+                if optimizer.converged():
+                    if show_progress_bar:
+                        print(f"\nConverged with loss: {np.round(mean_loss, 2)}")
+                    break
+        # Training finished:
+        self._trained_on = x
+
+        # Evaluate quality
+        if show_progress_bar:
+            try:
+                quality_control_fn, quality_control_msg = get_quality_metric(
+                    quality_controll_metric
+                )
+                metric = round(float(quality_control_fn(self)), 3)
+                print(f"Quality Score: {metric} " + quality_control_msg)
+            except Exception as e:
+                print(
+                    f"Quality controll did not work, we reset the variational \
+                        posterior,please check your setting! \
+                        \n Following error occured {e}"
+                )
+                self.train(
+                    learning_rate=learning_rate * 0.1,
+                    retrain_from_scratch=True,
+                    reset_optimizer=True,
+                )
```

Comment:
I don't get this. Does this reset? Doesn't it train `max_num_iters` epochs?

Reply:
Reset optimizer does just reset reinitialize the optimizer, thus e.g. resets the adaptive learning rate. Retrain from scratch does also reset the variational distribution. Yes, it does train max_num_iters as total (I did not call it epochs as there are no epochs).

Reply:
ahhh ok makes sense